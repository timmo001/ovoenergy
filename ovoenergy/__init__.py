"""Initialize the package."""
from __future__ import annotations
