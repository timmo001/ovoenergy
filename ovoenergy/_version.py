"""
Provides ovoenergy version information.
"""

# This file is auto-generated! Do not edit!
# Use `python -m incremental.update ovoenergy` to change this file.

from incremental import Version

__version__ = Version("ovoenergy", 1, 3, 0, dev=0)
__all__ = ["__version__"]
