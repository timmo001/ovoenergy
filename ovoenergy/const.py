"""Constants for the OVO Energy API client."""

# Base URLs
AUTH_BASE_URL = "https://my.ovoenergy.com/api/v2/auth"
SMARTPAY_BASE_URL = "https://smartpaymapi.ovoenergy.com"
KALUZA_BASE_URL = "https://api.eu1.prod.kaluza.com"

# Authentication endpoints
AUTH_LOGIN_URL = f"{AUTH_BASE_URL}/login"
AUTH_TOKEN_URL = f"{AUTH_BASE_URL}/token"

# Bootstrap endpoints
BOOTSTRAP_GRAPHQL_URL = f"{KALUZA_BASE_URL}/graphql/1"

# Usage endpoints
USAGE_DAILY_URL = f"{SMARTPAY_BASE_URL}/usage/api/daily"
USAGE_HALF_HOURLY_URL = f"{SMARTPAY_BASE_URL}/usage/api/half-hourly"


# Carbon endpoints
CARBON_FOOTPRINT_URL = f"{SMARTPAY_BASE_URL}/carbon-api"
CARBON_INTENSITY_URL = f"{SMARTPAY_BASE_URL}/carbon-bff/carbonintensity"

# GraphQL query for bootstrapping customer accounts
BOOTSTRAP_QUERY = """
query Bootstrap($customerId: ID!) {
  customer_nextV1(id: $customerId) {
    id
    customerAccountRelationships {
      edges {
        node {
          account {
            accountNo
            id
            accountSupplyPoints {
              ...AccountSupplyPoint
              __typename
            }
            __typename
          }
          __typename
        }
        __typename
      }
      __typename
    }
    __typename
  }
}

fragment AccountSupplyPoint on AccountSupplyPoint {
  startDate
  supplyPoint {
    sprn
    fuelType
    meterTechnicalDetails {
      meterSerialNumber
      mode
      type
      status
      __typename
    }
    address {
      addressLines
      postCode
      __typename
    }
    __typename
  }
  __typename
}
"""
