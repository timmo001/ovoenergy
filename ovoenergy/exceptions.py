"""Exceptions for the OVO Energy API."""


# Base Exceptions
class OVOEnergyException(Exception):
    """Base exception for OVO Energy."""


class OVOEnergyNoAccount(OVOEnergyException):
    """Exception for no account found."""


# API Exceptions
class OVOEnergyAPIException(OVOEnergyException):
    """Exception for API exceptions."""


class OVOEnergyAPINotAuthorized(OVOEnergyAPIException):
    """Exception for API client not authorized."""


class OVOEnergyAPINoCookies(OVOEnergyAPIException):
    """Exception for no cookies found."""
