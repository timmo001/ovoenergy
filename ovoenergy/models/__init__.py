"""OVO Energy: Models."""
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Extra, Field  # pylint: disable=no-name-in-module


class OVOBase(BaseModel):
    """Base class for all OVO models."""

    class Config:
        """Pydantic config."""

        extra = Extra.allow


class OVOInterval(OVOBase):
    """Interval model."""

    start: datetime = Field(None, alias="start")
    end: datetime = Field(None, alias="end")


class OVOMeterReadings(OVOBase):
    """Meter readings model."""

    start: float = Field(None, alias="start")
    end: float = Field(None, alias="end")


class OVOCost(OVOBase):
    """Cost model."""

    amount: Optional[float] = Field(None, alias="amount")
    currency_unit: Optional[str] = Field(None, alias="currencyUnit")


class OVODailyElectricity(OVOBase):
    """Daily electricity model."""

    consumption: Optional[float] = Field(None, alias="consumption")
    interval: Optional[OVOInterval] = Field(None, alias="interval")
    meter_readings: Optional[OVOMeterReadings] = Field(None, alias="meterReadings")
    has_half_hour_data: Optional[bool] = Field(None, alias="hasHalfHourData")
    cost: Optional[OVOCost] = Field(None, alias="cost")


class OVODailyGas(OVOBase):
    """Daily gas model."""

    consumption: Optional[float] = Field(None, alias="consumption")
    volume: Optional[float] = Field(None, alias="volume")
    interval: Optional[OVOInterval] = Field(None, alias="interval")
    meter_readings: Optional[OVOMeterReadings] = Field(None, alias="meterReadings")
    has_half_hour_data: Optional[bool] = Field(None, alias="hasHalfHourData")
    cost: Optional[OVOCost] = Field(None, alias="cost")


class OVOHalfHour(OVOBase):
    """Half hour model."""

    consumption: float = Field(None, alias="consumption")
    interval: OVOInterval = Field(None, alias="interval")
    unit: str = Field(None, alias="unit")


class OVODailyUsage(OVOBase):
    """Daily usage model."""

    electricity: Optional[list[OVODailyElectricity]] = Field(None, alias="electricity")
    gas: Optional[list[OVODailyGas]] = Field(None, alias="gas")


class OVOHalfHourUsage(OVOBase):
    """Half hour usage model."""

    electricity: Optional[list[OVOHalfHour]] = Field(None, alias="electricity")
    gas: Optional[list[OVOHalfHour]] = Field(None, alias="gas")


class OVOPlan(OVOBase):
    """Plan model."""

    standing_charge: Optional[float] = Field(None, alias="standingCharge")
    unit_rate: Optional[float] = Field(None, alias="unitRate")
    tariff: Optional[str] = Field(None, alias="tariff")
