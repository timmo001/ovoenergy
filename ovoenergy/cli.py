"""Enable CLI."""
import asyncio

import click

from ovoenergy.ovoenergy import OVOEnergy


@click.command()
@click.option("--username", "-u", help="Username")
@click.option("--password", "-p", help="Password")
@click.option("--account", "-a", help="Account Number")
@click.option("--date", "-D", help="Date")
@click.option("--daily", "-d", is_flag=True, help="Daily usage")
@click.option("--halfhour", "-h", is_flag=True, help="Half hourly usage")
def cli(username, password, account, date, daily=True, halfhour=False):
    """CLI for this package."""
    asyncio.run(handle(username, password, account, date, daily, halfhour))


async def handle(username, password, account, date, daily=True, halfhour=False) -> None:
    client = OVOEnergy()
    authenticated = await client.authenticate(username, password, account)
    print(f"Authenticated: {authenticated}")
    if authenticated:
        print("Authenticated.")
        if daily is True:
            usage = await client.get_daily_usage(date)
            if usage is not None:
                print("Usage:")
                print(usage)
                if usage.electricity is not None:
                    print("Electricity:")
                    count = 0
                    for x in usage.electricity:
                        count += 1
                        print(f"{count}.consumption: {x.consumption}")
                        print(f"{count}.interval.start: {x.interval.start}")
                        print(f"{count}.interval.end: {x.interval.end}")
                        # print(f"{count}.meter_readings.start: {x.meter_readings.start}")
                        # print(f"{count}.meter_readings.end: {x.meter_readings.end}")
                        print(f"{count}.has_hh_data: {x.has_half_hour_data}")
                        print(f"{count}.cost.amount: {x.cost.amount}")
                        print(f"{count}.cost.currency_unit: {x.cost.currency_unit}")
                if usage.gas is not None:
                    print("Gas:")
                    count = 0
                    for x in usage.gas:
                        count += 1
                        print(f"{count}.consumption: {x.consumption}")
                        print(f"{count}.volume: {x.volume}")
                        print(f"{count}.interval.start: {x.interval.start}")
                        print(f"{count}.interval.end: {x.interval.end}")
                        # print(f"{count}.meter_readings.start: {x.meter_readings.start}")
                        # print(f"{count}.meter_readings.end: {x.meter_readings.end}")
                        print(f"{count}.has_hh_data: {x.has_half_hour_data}")
                        print(f"{count}.cost.amount: {x.cost.amount}")
                        print(f"{count}.cost.currency_unit: {x.cost.currency_unit}")
        if halfhour is True:
            usage = await client.get_half_hourly_usage(date)
            if usage is not None:
                print("Usage:")
                print(usage)
                if usage.electricity is not None:
                    print("Electricity:")
                    count = 0
                    for x in usage.electricity:
                        count += 1
                        print(f"consumption: {x.consumption}")
                        print(f"{count}.interval.start: {x.interval.start}")
                        print(f"{count}.interval.end: {x.interval.end}")
                        print(f"{count}.unit: {x.unit}")
                if usage.gas is not None:
                    print("Gas:")
                    count = 0
                    for x in usage.gas:
                        count += 1
                        print(f"{count}.consumption: {x.consumption}")
                        print(f"{count}.interval.start: {x.interval.start}")
                        print(f"{count}.interval.end: {x.interval.end}")
                        print(f"{count}.unit: {x.unit}")


cli()  # pylint: disable=E1120
