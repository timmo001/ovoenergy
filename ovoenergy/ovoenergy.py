"""Get energy data from OVO's API."""
from datetime import datetime, timedelta

import aiohttp

from ovoenergy import (
    OVOCost,
    OVODailyElectricity,
    OVODailyGas,
    OVODailyUsage,
    OVOHalfHour,
    OVOHalfHourUsage,
    OVOInterval,
    OVOMeterReadings,
)


class OVOEnergy:
    """Class for OVOEnergy."""

    def __init__(self) -> None:
        """Initilalize."""
        self._account_id = None
        self._account_ids = None
        self._cookies = None
        self._customer_id = None
        self._username = None

    async def authenticate(self, username, password, account=None) -> bool:
        """Authenticate."""
        async with aiohttp.ClientSession() as session:
            response = await session.post(
                "https://my.ovoenergy.com/api/v2/auth/login",
                json={
                    "username": username,
                    "password": password,
                    "rememberMe": True,
                },
            )

            try:
                response.raise_for_status()
            except:
                pass

            if response.status is not 200:
                return False
            json_response = await response.json()

            if "code" in json_response and json_response["code"] == "Unknown":
                return False
            else:
                self._cookies = response.cookies
                response = await session.get(
                    "https://smartpaym.ovoenergy.com/api/customer-and-account-ids",
                    cookies=self._cookies,
                )
                json_response = await response.json()
                if "accountIds" in json_response:
                    self._account_ids = json_response["accountIds"]

                    if account:
                        if account in self._account_ids:
                            self._account_id = account
                        else:
                            return False
                    else:
                        self._account_id = self._account_ids[0]
                if "customerId" in json_response:
                    self._customer_id = json_response["customerId"]
                self._username = username

        return True

    async def get_daily_usage(self, month: str) -> OVODailyUsage:
        """Get daily usage data."""
        if month is None:
            return None
        electricity_usage = None
        gas_usage = None

        async with aiohttp.ClientSession() as session:
            response = await session.get(
                f"https://smartpaym.ovoenergy.com/api/energy-usage/daily/{self._account_id}?date={month}",
                cookies=self._cookies,
            )
            json_response = await response.json()

        if "electricity" in json_response:
            electricity = json_response["electricity"]
            if electricity and "data" in electricity:
                electricity_usage = []
                for usage in electricity["data"]:
                    if usage is not None:
                        electricity_usage.append(
                            OVODailyElectricity(
                                usage["consumption"]
                                if "consumption" in usage
                                else None,
                                OVOInterval(
                                    datetime.strptime(
                                        usage["interval"]["start"],
                                        "%Y-%m-%dT%H:%M:%S.%f",
                                    ),
                                    datetime.strptime(
                                        usage["interval"]["end"], "%Y-%m-%dT%H:%M:%S.%f"
                                    ),
                                )
                                if "interval" in usage
                                and "start" in usage["interval"]
                                and "end" in usage["interval"]
                                else None,
                                OVOMeterReadings(
                                    usage["meterReadings"]["start"],
                                    usage["meterReadings"]["end"],
                                )
                                if "meterReadings" in usage
                                else None,
                                usage["hasHhData"] if "hasHhData" in usage else None,
                                OVOCost(
                                    usage["cost"]["amount"],
                                    usage["cost"]["currencyUnit"],
                                )
                                if "cost" in usage
                                else None,
                            )
                        )

        if "gas" in json_response:
            gas = json_response["gas"]
            if gas and "data" in gas:
                gas_usage = []
                for usage in gas["data"]:
                    if usage is not None:
                        gas_usage.append(
                            OVODailyGas(
                                usage["consumption"]
                                if "consumption" in usage
                                else None,
                                usage["volume"] if "volume" in usage else None,
                                OVOInterval(
                                    datetime.strptime(
                                        usage["interval"]["start"],
                                        "%Y-%m-%dT%H:%M:%S.%f",
                                    ),
                                    datetime.strptime(
                                        usage["interval"]["end"], "%Y-%m-%dT%H:%M:%S.%f"
                                    ),
                                )
                                if "interval" in usage
                                and "start" in usage["interval"]
                                and "end" in usage["interval"]
                                else None,
                                OVOMeterReadings(
                                    usage["meterReadings"]["start"],
                                    usage["meterReadings"]["end"],
                                )
                                if "meterReadings" in usage
                                else None,
                                usage["hasHhData"] if "hasHhData" in usage else None,
                                OVOCost(
                                    usage["cost"]["amount"],
                                    usage["cost"]["currencyUnit"],
                                )
                                if "cost" in usage
                                else None,
                            )
                        )

        return OVODailyUsage(electricity_usage, gas_usage)

    async def get_half_hourly_usage(self, date: str) -> OVOHalfHourUsage:
        """Get half hourly usage data."""
        if date is None:
            return None
        electricity_usage = None
        gas_usage = None

        async with aiohttp.ClientSession() as session:
            response = await session.get(
                f"https://smartpaym.ovoenergy.com/api/energy-usage/half-hourly/{self._account_id}?date={date}",
                cookies=self._cookies,
            )
            json_response = await response.json()

        if "electricity" in json_response:
            electricity = json_response["electricity"]
            if electricity and "data" in electricity:
                electricity_usage = []
                for usage in electricity["data"]:
                    if usage is not None:
                        electricity_usage.append(
                            OVOHalfHour(
                                usage["consumption"]
                                if "consumption" in usage
                                else None,
                                OVOInterval(
                                    datetime.strptime(
                                        usage["interval"]["start"],
                                        "%Y-%m-%dT%H:%M:%S.%f",
                                    ),
                                    datetime.strptime(
                                        usage["interval"]["end"], "%Y-%m-%dT%H:%M:%S.%f"
                                    ),
                                )
                                if "interval" in usage
                                and "start" in usage["interval"]
                                and "end" in usage["interval"]
                                else None,
                                usage["unit"] if "unit" in usage else None,
                            )
                        )
        if "gas" in json_response:
            gas = json_response["gas"]
            if gas and "data" in gas:
                gas_usage = []
                for usage in gas["data"]:
                    if usage is not None:
                        gas_usage.append(
                            OVOHalfHour(
                                usage["consumption"]
                                if "consumption" in usage
                                else None,
                                OVOInterval(
                                    datetime.strptime(
                                        usage["interval"]["start"],
                                        "%Y-%m-%dT%H:%M:%S.%f",
                                    ),
                                    datetime.strptime(
                                        usage["interval"]["end"], "%Y-%m-%dT%H:%M:%S.%f"
                                    ),
                                )
                                if "interval" in usage
                                and "start" in usage["interval"]
                                and "end" in usage["interval"]
                                else None,
                                usage["unit"] if "unit" in usage else None,
                            )
                        )

        return OVOHalfHourUsage(electricity_usage, gas_usage)

    async def get_last_reading(self, date: datetime = None) -> OVOHalfHourUsage:
        date = date if date is not None else datetime.utcnow()
        print(f"DATE: {date}")
        usage: OVOHalfHourUsage = await self.get_half_hourly_usage(
            date.strftime("%Y-%m-%d")
        )
        if usage is None or usage.electricity is None or len(usage.electricity) == 0:
            date = date - timedelta(days=1)
            if date is datetime.utcnow() - timedelta(days=7):
                return None
            return await self.get_last_reading(date)
        return usage

    @property
    def account_id(self):
        return self._account_id

    @account_id.setter
    def account_id(self, new_account):
        if new_account in self._account_ids:
            self._account_id = new_account
        else:
            print("Invalid account ID")

    @property
    def account_ids(self):
        return self._account_ids

    @property
    def customer_id(self):
        return self._customer_id

    @property
    def username(self):
        return self._username
