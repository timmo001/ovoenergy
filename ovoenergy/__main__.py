"""OVO Energy: Main"""
from __future__ import annotations

import asyncio
from datetime import datetime, timedelta
from typing import Optional

import typer

from ._version import __version__
from .models import OVODailyUsage, OVOHalfHourUsage
from .models.carbon_intensity import OVOCarbonIntensity
from .models.footprint import OVOFootprint
from .models.plan import OVOPlan
from .ovoenergy import OVOEnergy

app = typer.Typer()
loop = asyncio.new_event_loop()
asyncio.set_event_loop(loop)


@app.command(name="daily", short_help="Get daily usage from OVO Energy")
def daily(
    username: str = typer.Option(..., help="OVO Energy username"),
    password: str = typer.Option(..., help="OVO Energy password"),
    account: str = typer.Option(None, help="OVO Energy account number"),
    date: str = typer.Option(None, help="Date to retrieve data for"),
) -> None:
    """Get daily usage from OVO Energy."""
    if date is None:
        # Get this month
        date = datetime.now().strftime("%Y-%m")

    ovo_usage: Optional[OVODailyUsage] = None

    client = OVOEnergy()

    authenticated = loop.run_until_complete(
        client.authenticate(username, password, account)
    )
    if authenticated:
        ovo_usage = loop.run_until_complete(client.get_daily_usage(date))

    typer.secho(
        ovo_usage.json() if ovo_usage is not None else '{"message": "No data"}',
        fg=typer.colors.GREEN,
    )


@app.command(name="halfhourly", short_help="Get half hourly usage from OVO Energy")
def half_hourly(
    username: str = typer.Option(..., help="OVO Energy username"),
    password: str = typer.Option(..., help="OVO Energy password"),
    account: str = typer.Option(None, help="OVO Energy account number"),
    date: str = typer.Option(None, help="Date to retrieve data for"),
) -> None:
    """Get half hourly usage from OVO Energy."""
    if date is None:
        # Get yesterday's date
        date = (datetime.now() - timedelta(days=1)).strftime("%Y-%m-%d")

    ovo_usage: Optional[OVOHalfHourUsage] = None

    client = OVOEnergy()
    authenticated = loop.run_until_complete(
        client.authenticate(username, password, account)
    )
    if authenticated:
        ovo_usage = loop.run_until_complete(client.get_half_hourly_usage(date))

    typer.secho(
        ovo_usage.json() if ovo_usage is not None else '{"message": "No data"}',
        fg=typer.colors.GREEN,
    )


@app.command(name="plan", short_help="Get plan from OVO Energy")
def plan(
    username: str = typer.Option(..., help="OVO Energy username"),
    password: str = typer.Option(..., help="OVO Energy password"),
    account: str = typer.Option(None, help="OVO Energy account number"),
) -> None:
    """Get rates from OVO Energy."""
    ovo_plan: Optional[OVOPlan] = None

    client = OVOEnergy()
    authenticated = loop.run_until_complete(
        client.authenticate(username, password, account)
    )
    if authenticated:
        ovo_plan = loop.run_until_complete(client.get_plan())

    typer.secho(
        ovo_plan.json() if ovo_plan is not None else '{"message": "No data"}',
        fg=typer.colors.GREEN,
    )


@app.command(name="carbon-footprint", short_help="Get carbon footprint from OVO Energy")
def carbon_footprint(
    username: str = typer.Option(..., help="OVO Energy username"),
    password: str = typer.Option(..., help="OVO Energy password"),
    account: str = typer.Option(None, help="OVO Energy account number"),
) -> None:
    """Get carbon footprint from OVO Energy."""
    ovo_footprint: Optional[OVOFootprint] = None

    client = OVOEnergy()
    authenticated = loop.run_until_complete(
        client.authenticate(username, password, account)
    )
    if authenticated:
        ovo_footprint = loop.run_until_complete(client.get_footprint())

    typer.secho(
        ovo_footprint.json() if ovo_footprint is not None else '{"message": "No data"}',
        fg=typer.colors.GREEN,
    )


@app.command(name="carbon-intensity", short_help="Get carbon intensity from OVO Energy")
def carbon_intensity(
    username: str = typer.Option(..., help="OVO Energy username"),
    password: str = typer.Option(..., help="OVO Energy password"),
    account: str = typer.Option(None, help="OVO Energy account number"),
) -> None:
    """Get carbon intensity from OVO Energy."""
    ovo_carbon_intensity: Optional[OVOCarbonIntensity] = None

    client = OVOEnergy()
    authenticated = loop.run_until_complete(
        client.authenticate(username, password, account)
    )
    if authenticated:
        ovo_carbon_intensity = loop.run_until_complete(client.get_carbon_intensity())

    typer.secho(
        ovo_carbon_intensity.json()
        if ovo_carbon_intensity is not None
        else '{"message": "No data"}',
        fg=typer.colors.GREEN,
    )


@app.command(name="version", short_help="Module Version")
def version() -> None:
    """Module Version"""
    typer.secho(__version__.public(), fg=typer.colors.CYAN)


if __name__ == "__main__":
    app()
