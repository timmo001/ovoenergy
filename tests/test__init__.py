"""Tests for the client module."""

from datetime import datetime, timedelta

from aioresponses import aioresponses
import pytest
from syrupy.assertion import SnapshotAssertion

from ovoenergy import OVOEnergy
from ovoenergy.exceptions import (
    OVOEnergyAPINoCookies,
    OVOEnergyAPINotAuthorized,
    OVOEnergyNoAccount,
)

from . import ACCOUNT, ACCOUNT_BAD, PASSWORD, RESPONSE_JSON_AUTH, USERNAME


@pytest.mark.asyncio
async def test_authorize(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
    snapshot: SnapshotAssertion,
) -> None:
    """Test authorize."""
    await ovoenergy_client.authenticate(USERNAME, PASSWORD)

    assert not ovoenergy_client.oauth_expired

    assert ovoenergy_client.oauth is not None

    assert ovoenergy_client.oauth.access_token == snapshot(
        name="authorize_oauth_access_token",
    )
    assert ovoenergy_client.oauth.expires_in == snapshot(
        name="authorize_oauth_expires_in",
    )
    assert ovoenergy_client.oauth.refresh_expires_in == snapshot(
        name="authorize_oauth_refresh_expires_in",
    )
    assert ovoenergy_client.account_id == snapshot(
        name="authorize_account_id",
    )
    assert ovoenergy_client.account_ids == snapshot(
        name="authorize_account_id",
    )
    assert ovoenergy_client.customer_id == snapshot(
        name="authorize_customer_id",
    )
    assert ovoenergy_client.username == snapshot(
        name="authorize_username",
    )


@pytest.mark.asyncio
async def test_bootstrap(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
    snapshot: SnapshotAssertion,
) -> None:
    """Test bootstrap."""
    await ovoenergy_client.authenticate(USERNAME, PASSWORD)

    assert await ovoenergy_client.bootstrap_accounts() == snapshot(
        name="bootstrap_accounts",
    )


@pytest.mark.asyncio
async def test_get_daily_usage(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
    snapshot: SnapshotAssertion,
) -> None:
    """Test get daily usage."""
    with pytest.raises(OVOEnergyNoAccount):
        await ovoenergy_client.get_daily_usage("2024-01")

    await ovoenergy_client.authenticate(USERNAME, PASSWORD)

    await ovoenergy_client.bootstrap_accounts()

    assert await ovoenergy_client.get_daily_usage("2024-01") == snapshot(
        name="daily_usage",
    )


@pytest.mark.asyncio
async def test_get_half_hourly_usage(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
    snapshot: SnapshotAssertion,
) -> None:
    """Test get half hourly usage."""
    with pytest.raises(OVOEnergyNoAccount):
        await ovoenergy_client.get_half_hourly_usage("2024-01-01")

    await ovoenergy_client.authenticate(USERNAME, PASSWORD)

    await ovoenergy_client.bootstrap_accounts()

    assert await ovoenergy_client.get_half_hourly_usage("2024-01-01") == snapshot(
        name="half_hourly_usage",
    )


@pytest.mark.asyncio
async def test_get_plans(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
    snapshot: SnapshotAssertion,
) -> None:
    """Test get plans."""
    with pytest.raises(OVOEnergyNoAccount):
        await ovoenergy_client.get_plans()

    await ovoenergy_client.authenticate(USERNAME, PASSWORD)

    await ovoenergy_client.bootstrap_accounts()

    assert await ovoenergy_client.get_plans() == snapshot(
        name="plans",
    )


@pytest.mark.asyncio
async def test_get_footprint(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
    snapshot: SnapshotAssertion,
) -> None:
    """Test get footprint."""
    with pytest.raises(OVOEnergyNoAccount):
        await ovoenergy_client.get_footprint()

    await ovoenergy_client.authenticate(USERNAME, PASSWORD)

    await ovoenergy_client.bootstrap_accounts()

    assert await ovoenergy_client.get_footprint() == snapshot(
        name="footprint",
    )


@pytest.mark.asyncio
async def test_get_carbon_intensity(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
    snapshot: SnapshotAssertion,
) -> None:
    """Test get carbon intensity."""
    await ovoenergy_client.authenticate(USERNAME, PASSWORD)

    assert await ovoenergy_client.get_carbon_intensity() == snapshot(
        name="carbon_intensity",
    )


@pytest.mark.asyncio
async def test_bootstrap_custom_account(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
    snapshot: SnapshotAssertion,
) -> None:
    """Test bootstrap custom account."""
    await ovoenergy_client.authenticate(USERNAME, PASSWORD)

    ovoenergy_client.custom_account_id = ACCOUNT

    await ovoenergy_client.bootstrap_accounts()

    assert await ovoenergy_client.get_daily_usage("2024-01") == snapshot(
        name="bootstrap_accounts_custom_account",
    )


@pytest.mark.asyncio
async def test_bad_account(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
) -> None:
    """Test bad account."""
    await ovoenergy_client.authenticate(USERNAME, PASSWORD)

    await ovoenergy_client.bootstrap_accounts()

    ovoenergy_client.custom_account_id = ACCOUNT_BAD

    with pytest.raises(OVOEnergyNoAccount):
        await ovoenergy_client.get_daily_usage("2024-01")


# pylint: disable=protected-access
@pytest.mark.asyncio
async def test_no_cookies(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
) -> None:
    """Test no cookies."""
    await ovoenergy_client.authenticate(USERNAME, PASSWORD)

    await ovoenergy_client.bootstrap_accounts()

    ovoenergy_client._cookies = None

    with pytest.raises(OVOEnergyAPINoCookies):
        await ovoenergy_client.get_daily_usage("2024-01")


# pylint: disable=protected-access
@pytest.mark.asyncio
async def test_no_auth(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
) -> None:
    """Test no auth."""
    await ovoenergy_client.authenticate(USERNAME, PASSWORD)

    await ovoenergy_client.bootstrap_accounts()

    ovoenergy_client._oauth = None

    with pytest.raises(OVOEnergyAPINotAuthorized):
        await ovoenergy_client.get_daily_usage("2024-01")


# pylint: disable=protected-access
@pytest.mark.asyncio
async def test_oauth_expired(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
) -> None:
    """Test oauth expired."""
    an_hour_ago = datetime.now() - timedelta(hours=1)

    await ovoenergy_client.authenticate(USERNAME, PASSWORD)

    await ovoenergy_client.bootstrap_accounts()

    assert ovoenergy_client._oauth is not None

    ovoenergy_client._oauth.expires_at = an_hour_ago

    await ovoenergy_client.get_daily_usage("2024-01")

    ovoenergy_client._oauth.expires_at = an_hour_ago

    mock_aioresponse.clear()
    mock_aioresponse.get(
        "https://my.ovoenergy.com/api/v2/auth/token",
        status=403,
        repeat=True,
    )

    with pytest.raises(OVOEnergyAPINotAuthorized):
        await ovoenergy_client.get_daily_usage("2024-01")


@pytest.mark.asyncio
async def test_forbidden(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
) -> None:
    """Test forbidden."""
    await ovoenergy_client.authenticate(USERNAME, PASSWORD)

    await ovoenergy_client.bootstrap_accounts()

    mock_aioresponse.clear()
    mock_aioresponse.get(
        f"https://smartpaymapi.ovoenergy.com/usage/api/daily/{ACCOUNT}?date=2024-01",
        status=403,
        repeat=True,
    )

    with pytest.raises(OVOEnergyAPINotAuthorized):
        await ovoenergy_client.get_daily_usage("2024-01")


@pytest.mark.asyncio
async def test_auth_not_found(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
) -> None:
    """Test auth endpoint not found."""
    mock_aioresponse.clear()
    mock_aioresponse.post(
        "https://my.ovoenergy.com/api/v2/auth/login",
        status=404,
        repeat=True,
    )

    assert not await ovoenergy_client.authenticate(USERNAME, PASSWORD)


@pytest.mark.asyncio
async def test_auth_code_not_found(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
) -> None:
    """Test auth endpoint code not found."""
    mock_aioresponse.clear()
    mock_aioresponse.post(
        "https://my.ovoenergy.com/api/v2/auth/login",
        status=204,
        repeat=True,
    )

    assert not await ovoenergy_client.authenticate(USERNAME, PASSWORD)


@pytest.mark.asyncio
async def test_auth_code_unknown(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
) -> None:
    """Test auth token not found."""
    mock_aioresponse.clear()
    mock_aioresponse.post(
        "https://my.ovoenergy.com/api/v2/auth/login",
        status=200,
        payload={"code": "Unknown"},
        repeat=True,
    )

    assert not await ovoenergy_client.authenticate(USERNAME, PASSWORD)


@pytest.mark.asyncio
async def test_auth_token_not_found(
    ovoenergy_client: OVOEnergy,
    mock_aioresponse: aioresponses,
) -> None:
    """Test auth token not found."""
    mock_aioresponse.clear()
    mock_aioresponse.post(
        "https://my.ovoenergy.com/api/v2/auth/login",
        payload=RESPONSE_JSON_AUTH,
        status=200,
        repeat=True,
    )
    mock_aioresponse.get(
        "https://my.ovoenergy.com/api/v2/auth/token",
        status=404,
        repeat=True,
    )

    assert not await ovoenergy_client.authenticate(USERNAME, PASSWORD)
