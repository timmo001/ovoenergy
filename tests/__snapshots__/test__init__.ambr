# serializer version: 1
# name: test_authorize[authorize_account_id]
  None
# ---
# name: test_authorize[authorize_customer_id]
  None
# ---
# name: test_authorize[authorize_oauth_access_token]
  'test'
# ---
# name: test_authorize[authorize_oauth_expires_in]
  3600
# ---
# name: test_authorize[authorize_oauth_refresh_expires_in]
  0
# ---
# name: test_authorize[authorize_username]
  'test'
# ---
# name: test_bootstrap[bootstrap_accounts]
  BootstrapAccounts(account_ids=[123456789], customer_id=UUID('5cafe9c4-a942-46b5-a67c-5882eba0a03c'), selected_account_id=123456789, accounts=[Account(account_id=123456789, is_payg=False, is_blocked=False, supplies=[Supply(mpxn='3456766576', fuel='gas', is_onboarding=False, start=datetime.datetime(2024, 1, 1, 23, 0, tzinfo=datetime.timezone.utc), is_payg=False, supply_point_info=SupplyPointInfo(meter_type='AB123', meter_not_found=False, address=['ADDR'])), Supply(mpxn='4536756746', fuel='electricity', is_onboarding=False, start=datetime.datetime(2024, 1, 1, 23, 0, tzinfo=datetime.timezone.utc), is_payg=False, supply_point_info=SupplyPointInfo(meter_type='AB123', meter_not_found=False, address=['ADDR']))])], is_first_login=False)
# ---
# name: test_bootstrap_custom_account[bootstrap_accounts_custom_account]
  OVODailyUsage(electricity=None, gas=None)
# ---
# name: test_get_carbon_intensity[carbon_intensity]
  OVOCarbonIntensity(forecast=[OVOCarbonIntensityForecast(time_from='2pm', intensity=82, level='low', colour='#0A9928', colour_v2='#0D8426')], current='low', greentime=None)
# ---
# name: test_get_daily_usage[daily_usage]
  OVODailyUsage(electricity=None, gas=None)
# ---
# name: test_get_footprint[footprint]
  OVOFootprint(from_='2024-01-01T00:00:00Z', to='2024-01-01T23:59:59.999000Z', carbon_reduction_product_ids=[], carbon_footprint=OVOCarbonFootprint(carbon_kg=2200.1234, carbon_saved_kg=0.0, k_wh=1578.3246, breakdown=OVOFootprintBreakdown(electricity=OVOFootprintElectricity(carbon_kg=200.1234, carbon_saved_kg=230.02, k_wh=65645.92), gas=OVOFootprintGas(carbon_kg=2000.1234, carbon_saved_kg=340.02, k_wh=10664.74363579))))
# ---
# name: test_get_half_hourly_usage[half_hourly_usage]
  OVOHalfHourUsage(electricity=None, gas=None)
# ---
# name: test_get_plans[plans]
  OVOPlans(electricity=[], gas=[])
# ---
